### PR TITLE
Template Discovery: added support for 2nd search query

### PR DESCRIPTION
**Pull request by vlada-shubina**
Template discovery tool: added support for searching packs with `packageType` equal to `Template`

### CODE REVIEW

File: src/Microsoft.TemplateSearch.TemplateDiscovery/Properties/launchSettings.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "Microsoft.TemplateSearch.TemplateDiscovery": {
+      "commandName": "Project",
+      "commandLineArgs": "--onePage --basePath test6 --allowPreviewPacks -v --savePacks "
+    }
+  }
+}
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/Properties/launchSettings.json:8]: Is this change intentional?

Comment (vlada-shubina) [reply src/Microsoft.TemplateSearch.TemplateDiscovery/Properties/launchSettings.json:8]: No, it's not. Fixed it.

File: src/Microsoft.TemplateSearch.TemplateDiscovery/Nuget/NugetPackProvider.cs [hunk 2]
```diff
@@ -110,19 +95,26 @@ private async Task<string> DownloadPackageAsync(NugetPackageSourceInfo packinfo)
...
+                        TotalDownloads = packinfo.TotalDownloads
+
+                    };
                 }
             }
-            catch
+            catch (Exception e)
             {
-                Console.WriteLine($"Failed to download package {packinfo.Id} {packinfo.Version}");
+                Console.WriteLine($"Failed to download package {packinfo.Id} {packinfo.Version}, reason: {e.Message}.");
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/Nuget/NugetPackProvider.cs:110]: Since this is used internally, maybe do `e.ToString()` to include whole stack.

File: src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs [hunk 3]
```diff
@@ -30,65 +30,89 @@ public PackSourceChecker(IPackProvider packProvider, PackPreFilterer packPreFilt
...
-                {
-                    Verbose.WriteLine($"{packInfo.Id}::{packInfo.Version} is skipped, {preFilterResult.PreFilterResults.Reason}");
-                    checkResultList.Add(preFilterResult);
-                }
-                else
+                Console.WriteLine($"Processing pack provider {packProvider.Name}:");
+                int count = 0;
+                Console.WriteLine($"{await packProvider.GetPackageCountAsync().ConfigureAwait(false)} packs discovered, starting processing");
+
+                await foreach (IPackInfo sourceInfo in packProvider.GetCandidatePacksAsync().ConfigureAwait(false))
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:42]: Woot, `await foreach` I need to brush my C# skills it seems, never seen this...
Ignore this comment :)

File: src/Microsoft.TemplateSearch.TemplateDiscovery/Program.cs [hunk 4]
```diff
@@ -143,6 +146,27 @@ private static bool TryParseArgs(string[] args, ScraperConfig config)
                     Verbose.IsEnabled = true;
                     ++index;
                 }
+                else if (string.Equals(args[index], _providers, StringComparison.Ordinal))
+                {
+                    if (TryGetFlagValue(args, index, out string providersList))
+                    {
+                        config.Providers.AddRange(providersList.Split('|', StringSplitOptions.TrimEntries));
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/Program.cs:153]: `StringSplitOptions.TrimEntries` woot :D
Ignore this comment :)

**Review by DavidKarlas — APPROVED** [review]
If you feel like addressing this issues, fine otherwise can be merged as is

File: src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs [hunk 5]
```diff
@@ -30,65 +30,89 @@ public PackSourceChecker(IPackProvider packProvider, PackPreFilterer packPreFilt
...
-                PackCheckResult preFilterResult = PrefilterPackInfo(packInfo);
-                if (preFilterResult.PreFilterResults.ShouldBeFiltered)
-                {
-                    Verbose.WriteLine($"{packInfo.Id}::{packInfo.Version} is skipped, {preFilterResult.PreFilterResults.Reason}");
-                    checkResultList.Add(preFilterResult);
-                }
-                else
+                Console.WriteLine($"Processing pack provider {packProvider.Name}:");
+                int count = 0;
+                Console.WriteLine($"{await packProvider.GetPackageCountAsync().ConfigureAwait(false)} packs discovered, starting processing");
```
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:40]: Are we only doing this to print the number to the screen? An additional network call for this seems a bit expensive. Maybe we could await on `packProvider.GetCandidatePacksAsync()` and use the result for both printing the count and downloading the package content?

File: src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs [hunk 6]
```diff
@@ -30,65 +30,89 @@ public PackSourceChecker(IPackProvider packProvider, PackPreFilterer packPreFilt
         public async Task<PackSourceCheckResult> CheckPackagesAsync()
         {
             List<PackCheckResult> checkResultList = new List<PackCheckResult>();
+            HashSet<IPackInfo> alreadySeenPacks = new HashSet<IPackInfo>();
```
Comment (bozturkMSFT) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:33]: We learn the number of packs early on. It may be good to use that number to pre-allocate some capacity for this when instantiating (also applies to `checkResultList`, right?).

File: src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs [hunk 7]
```diff
@@ -30,65 +30,89 @@ public PackSourceChecker(IPackProvider packProvider, PackPreFilterer packPreFilt
...
-                    PackCheckResult packCheckResult = _packChecker.TryGetTemplatesInPack(packInfo, _additionalDataProducers, alreadySeenTemplateIdentities);
-                    checkResultList.Add(packCheckResult);
+                    if (alreadySeenPacks.Contains(sourceInfo))
+                    {
+                        Verbose.WriteLine($"Package {sourceInfo.Id}::{sourceInfo.Version} is already processed.");
+                        continue;
+                    }
+                    alreadySeenPacks.Add(sourceInfo);
 
+                    IDownloadedPackInfo packInfo = await packProvider.DownloadPackageAsync(sourceInfo).ConfigureAwait(false);
```
Comment (bozturkMSFT) [inline src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:51]: As far as I know, NuGet client allows downloading directly into memory without writing onto disk. If we are planning to delete the packages anyway, could we do this all in memory?

**Review by bozturkMSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:40]: First time I've used the tool, it was endless process and you didn't know how many template packs are left. For me as tool user having this number printed worth waiting the time

[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs:40]: For CancellationToken and other improvements below - they are valid however tbh I cannot justify the time spending on this tool comparing to other tasks, or even create issue for it. It's internal tool and we use it manually 2 times per months. We might want to consider improvements when will start automating runs.

File: src/Microsoft.TemplateSearch.TemplateDiscovery/Nuget/NugetPackScraper.cs [hunk 8]
```diff
@@ -10,11 +11,31 @@ namespace Microsoft.TemplateSearch.TemplateDiscovery.Nuget
...
+
+            if (!config.Providers.Any())
+            {
+                providers.AddRange(SupportedProviders.Select(kvp => new NugetPackProvider(kvp.Key, kvp.Value, config.BasePath, config.PageSize, config.RunOnlyOnePage, config.IncludePreviewPacks)));
+            }
+            else
+            {
+                foreach (string provider in config.Providers.Distinct(StringComparer.OrdinalIgnoreCase))
+                {
+                    providers.Add(new NugetPackProvider(provider, SupportedProviders[provider], config.BasePath, config.PageSize, config.RunOnlyOnePage, config.IncludePreviewPacks));
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateSearch.TemplateDiscovery/Nuget/NugetPackScraper.cs:34]: Providers are validated in Program.cs when parsed so it's not possible the provider which does not exist in SupportedProviders. I believe for the background and purpose of the tool it's sufficient handling.

**Review by bozturkMSFT — APPROVED** [review]
